vindex/cmd/logandmap/main.go

```diff
@@ -0,0 +1,331 @@
+}
+
+// maintainMap reads entries from the log and sync them to the vindex.
+func maintainMap(ctx context.Context, vi *vindex.VerifiableIndex) {
+	ticker := time.NewTicker(10 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			if err := vi.Update(ctx); err != nil {
+				klog.Warning(err)
+			}
+		}
+	}
+}
+
+// submitEntries continually creates new log entries and submits them to the log.
+// Entries should be json-encoded LogEntrys.
+// The module should be randomly pulled from a list of [foo, bar, baz, splat]
+// The version should be the current timestamp, as a string
+// The hash should be the sha256 of the module+version
+func submitEntries(ctx context.Context, appender *tessera.Appender) {
+	modules := []string{"foo", "bar", "baz", "splat"}
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			klog.Info("Context cancelled, stopping log appender")
+			return
+		case <-ticker.C:
+			module := modules[r.Intn(len(modules))]
+			version := time.Now().Format(time.RFC3339Nano)
+			h := sha256.Sum256([]byte(module + version))
+			entry := LogEntry{
+				Module:  module,
+				Version: version,
+				Hash:    h[:],
+			}
+			data, err := json.Marshal(entry)
+			if err != nil {
+				klog.Errorf("Failed to marshal log entry: %v", err)
+				continue
+			}
+			if idx, err := appender.Add(ctx, tessera.NewEntry(data))(); err != nil {
+				klog.Errorf("Failed to append to log: %v", err)
+			} else {
+				klog.Infof("Appended entry for %s@%s at index %d", module, version, idx.Index)
+			}
+		}
+	}
+}
+
+func runWebServer(vi *vindex.VerifiableIndex) {
+	web := NewServer(func(s string) string {
+		idxes := vi.Lookup(s)
+		return fmt.Sprintf("Indices in log: %v", idxes)
+	})
+
+	fs := http.FileServer(http.Dir(*posixLogDir))
+	r := mux.NewRouter()
+	r.PathPrefix("/log/").Handler(http.StripPrefix("/log/", fs))
+	web.registerHandlers(r)
+	hServer := &http.Server{
+		Addr:    *addr,
+		Handler: r,
+	}
+	go func() {
+		_ = hServer.ListenAndServe()
+	}()
+	klog.Infof("Started HTTP server listening on %s", *addr)
+}
+
+// Read log private key from file or environment variable and generate the
+// note Signer and Verifier pair for it.
+func getSignerVerifierOrDie() (note.Signer, note.Verifier) {
+	var privKey string
+	var err error
+	if len(*privKeyFile) > 0 {
+		privKey, err = getKeyFile(*privKeyFile)
+		if err != nil {
+			klog.Exitf("Unable to get private key: %v", err)
+		}
+	} else {
+		privKey = os.Getenv("LOG_PRIVATE_KEY")
+		if len(privKey) == 0 {
+			klog.Exit("Supply private key file path using --private_key or set LOG_PRIVATE_KEY environment variable")
+		}
+	}
+	s, v, err := signerVerifierFromSkey(privKey)
+	if err != nil {
+		klog.Exitf("Failed to get signer/verifier: %v", err)
+	}
+	return s, v
+}
+
+func signerVerifierFromSkey(skey string) (note.Signer, note.Verifier, error) {
```

Comment:
I've seen variations on this a few times now (e.g. there's also https://github.com/FiloSottile/torchwood/blob/v0.5.0/note.go#L17) - perhaps the pub-from-priv bit of this is worth "promoting" into `formats`?